### PR TITLE
Script for the Clustered Feature Importance Algorithm

### PR DESCRIPTION
**Pull request by Harkishan-99**
Added the Clustered Feature Importance Algorithm, a modified version of MDA (mean decreased accuracy) algorithm. It is robust to both linear and non-linear substitution effect. It is discussed in de Prados presentation slide (https://papers.ssrn.com/sol3/papers.cfm?abstract_id=3517595).

### CODE REVIEW

File: mlfinlab/feature_importance/cfi.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    for col0 in features_cols:
+        vi = []
+        for col1 in features_cols:
+            x = df[col0].values
+            y = df[col1].values
+            vi.append(variation_of_information_score(x,y,n_bins=None,normalize=True))
+        VI[col0] = vi
+    return VI
+
+def cluster_features(X : pd.DataFrame, C : int = None):
```
Comment (proskurin) [inline mlfinlab/feature_importance/cfi.py]: Can we change C parameter name to n_clusters as it defined in clustered_feature_importance?

File: mlfinlab/feature_importance/cfi.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+#Imports
+import numpy as np
+import pandas as pd
+import mlfinlab as ml
+from sklearn.metrics import log_loss
+from scipy.cluster.hierarchy import linkage, fcluster
+from mlfinlab.codependence import variation_of_information_score
+
+def get_VI_matrix(df : pd.DataFrame):
```
Comment (proskurin) [inline mlfinlab/feature_importance/cfi.py]: Can we make a universal function which builds dependence matrix using functions from mlfinlab.codependence module?

File: mlfinlab/feature_importance/cfi.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    (param) C : (int) number of clusters to form.
+                      Must be less the total number of features.
+                      If None then it returns optimal number of clusters
+                      decided by the ONC Algorithm.
+
+    (return) : (array) of feature subsets
+    '''
+
+    vi_matrix =  get_VI_matrix(X)
+    link = linkage(vi_matrix, 'single')
```
Comment (proskurin) [inline mlfinlab/feature_importance/cfi.py]: Set's also make linkage as parameter. What I've seen from my practice 'ward' linkage works much better, so let the user decide what linkage he wants to use.

Comment (proskurin) [inline mlfinlab/feature_importance/cfi.py]: As far as I understand we should apply linkage to distance matrix rather than dependence. So in this case we calculate dependence matrix  (vi_matrix) -> distance, for example angular (or abs angular or squared angular) can also be a function parameter on dependence matrix -> linkage(dependence matrix)

File: mlfinlab/feature_importance/cfi.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    if C is not None:
+        if C >= len(X.columns):
+            raise ValueError('Number of Clusters Must be less than the number of features')
+        clusters = fcluster(link, t = C, criterion='maxclust')
+        clustered_subsets = [[f for c, f in zip(clusters, X.columns) if c == ci] for ci in range(1, C+1)]
+        return clustered_subsets
+
+    else:
+        C = len(ml.clustering.get_onc_clusters(vi_matrix)[1])
```
Comment (proskurin) [inline mlfinlab/feature_importance/cfi.py]: Regarding the comment above , in this case ONC is used correctly cause it is applied directly to dependence matrix instead of distance, great work!

File: mlfinlab/feature_importance/cfi.py (hunk 3 above)
Comment (Harkishan-99) [reply mlfinlab/feature_importance/cfi.py]: As per Dr de Prado, he mentioned only the use of a single linkage. Below I have attached a slide from the original presentation (highlighting the same). As per my understanding, he used a single linkage because it can address the issue of multicollinearity (presence of redundant features).
cfi

File: mlfinlab/feature_importance/cfi.py (hunk 1 above)
Comment (Harkishan-99) [reply mlfinlab/feature_importance/cfi.py]: I agree that makes total sense. I kept it so because it was just used as a helper function.

File: mlfinlab/feature_importance/cfi.py (hunk 3 above)
Comment (proskurin) [reply mlfinlab/feature_importance/cfi.py]: Of course, de Prado also uses single-linkage algorithm in HPR. However, on practice user can deviate from presentation materials and apply other linkage functions

File: mlfinlab/feature_importance/cfi.py (hunk 2 above)
Comment (Harkishan-99) [reply mlfinlab/feature_importance/cfi.py]: Can you elaborate on what you meant by  universal function?

File: mlfinlab/feature_importance/cfi.py (hunk 3 above)
Comment (Harkishan-99) [reply mlfinlab/feature_importance/cfi.py]: Okay, I will make it a parameter and set the default to single.

File: mlfinlab/feature_importance/cfi.py (hunk 2 above)
Comment (proskurin) [reply mlfinlab/feature_importance/cfi.py]: So, as of now in our codependence module we have:
- VI
- Mutual info
- Distance correlation

And simple linear correlation. It would be greate to have a function:

get_dependence_function(algorithm='variation_information') where method is one of algorithms described above or just Pearson correlation

Comment (Harkishan-99) [reply mlfinlab/feature_importance/cfi.py]: Yeah, that would be very handy. I will try to do that.

File: mlfinlab/feature_importance/cfi.py (hunk 3 above)
Comment (Harkishan-99) [reply mlfinlab/feature_importance/cfi.py]: I get it but the distance matrix can be applied during the presence of linear substitution effect, but the VI matrix is effective in presence of both linear and non-linear substitution effects. I really want to know your @proskurin opinion on these.

Comment (proskurin) [reply mlfinlab/feature_importance/cfi.py]: As far as I understand distance matrix is always an input for clustering. If for example we have negative values (for linear correlation for example), clustering algorithm will raise error. So VI is a measure of dependence which always needs to be transformed  into distance matrix. As far as I remember, Marcos stated that we need to apply angular distance operator before clustering

Comment (Harkishan-99) [reply mlfinlab/feature_importance/cfi.py]: Okay, as you mentioned to add a universal function (in the above comments), I would create a new script in the codependence section that can get both dependence and distance matrix.

File: mlfinlab/feature_importance/cfi.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+from sklearn.metrics import log_loss
+from scipy.cluster.hierarchy import linkage, fcluster
+from information_matrix import  get_dependence_matrix , get_distance_matrix
+
+
+def cluster_features(X : pd.DataFrame , linkage_method : str , n_clusters : int = None ):
+    '''
+    Get Clusters containing features subsets
+
+    (param) X : (pd.DataFrame) of features
```
Comment (Jackal08) [inline mlfinlab/feature_importance/cfi.py]: This doc string looks a bit strange. 

1. has `(param)`
2. check argument names

File: mlfinlab/feature_importance/cfi.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+import pandas as pd
+import mlfinlab as ml
+from sklearn.metrics import log_loss
+from scipy.cluster.hierarchy import linkage, fcluster
+from information_matrix import  get_dependence_matrix , get_distance_matrix
+
+
+def cluster_features(X : pd.DataFrame , linkage_method : str , n_clusters : int = None ):
+    '''
+    Get Clusters containing features subsets
```
Comment (Jackal08) [inline mlfinlab/feature_importance/cfi.py]: Lets include way more details in these doc strings. This part of the code is almost the most important as it tells people what it does and how it is used.

File: mlfinlab/feature_importance/cfi.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    (param) linkage_method : (str) method of linkage to be used for clustering. Methods include -
+                                  'single' , 'ward' , 'complete' , 'average' , 'weighted' and 'centroid'.
+    (param) C : (int) number of clusters to form.
+                      Must be less the total number of features.
+                      If None then it returns optimal number of clusters
+                      decided by the ONC Algorithm.
+
+    (return) : (array) of feature subsets
+    '''
+    vi_matrix = get_dependence_matrix(df =X , base_algorithm = 'variation_of_information')
```
Comment (Jackal08) [inline mlfinlab/feature_importance/cfi.py]: Please add inline comments describing the various lines of code.

File: mlfinlab/feature_importance/cfi.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+            fold_metrics_values.loc[i] = -scoring(y.iloc[test], prob, sample_weight=sample_weight[test],
+                                                  labels=model.classes_)
+        else:
+            fold_metrics_values.loc[i] = scoring(y.iloc[test], pred, sample_weight=sample_weight[test])
+
+        # Get feature specific metric on out-of-sample fold
+        for j in clustered_subsets: #Instead of Individual Features , clustered subsets are used
+            X1_ = X.iloc[test, :].copy(deep=True)
+
+            for j_i in j:
```
Comment (Jackal08) [inline mlfinlab/feature_importance/cfi.py]: A lot more inline comments please.

File: test [hunk 9]
```diff
@@ -0,0 +1 @@
+mlfinlab/feature_importance/cfi.py
```
Comment (Jackal08) [inline test]: Delete this file.

File: mlfinlab/feature_importance/__init__.py [hunk 10]
```diff
@@ -8,3 +8,5 @@
 from mlfinlab.feature_importance.orthogonal import (feature_pca_analysis, get_pca_rank_weighted_kendall_tau,
                                                     get_orthogonal_features)
 from mlfinlab.feature_importance.fingerpint import RegressionModelFingerprint, ClassificationModelFingerprint
+
```
Comment (Jackal08) [inline mlfinlab/feature_importance/__init__.py]: Remove white space

File: mlfinlab/codependence/information_matrix.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+import pandas as pd
+
+from mlfinlab.codependence.information import variation_of_information_score,mutual_info_score
+from mlfinlab.codependence.correlation import angular_distance,absolute_angular_distance,\
+                                              squared_angular_distance,distance_correlation
+
+
+def get_dependence_matrix(df : pd.DataFrame, base_algorithm : str ,n_bins : int = None, normalize : bool = True):
+    """
+    This function returns dependence and distance matrix for the methods of Information Codependence.
```
Comment (Jackal08) [inline mlfinlab/codependence/information_matrix.py]: Does it get both the dependence and the distance? I see there is another function titled get distance.

File: mlfinlab/codependence/information_matrix.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+            x = df[col0].values
+            y = df[col1].values
+            vi.append(algo(x,y,n_bins,normalize))
+        info_mtx[col0] = vi
+    return info_mtx
+
+def get_distance_matrix(X : pd.DataFrame or np.array, transformation : str = 'angular'):
+
+    """
+    This function returns the distance_matrix for the given metrix X.
```
Comment (Jackal08) [inline mlfinlab/codependence/information_matrix.py]: Spelling.

File: mlfinlab/codependence/information_matrix.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
+"""
```
Comment (Jackal08) [inline mlfinlab/codependence/information_matrix.py]: @proskurin do we need a separate file for this code? Should this code form part of the codependence module, will users make use of these two functions in areas other than just clustrered feature importance? 

I think we need to provide a lot more documentation to show how these functions are used. If they are just for feature clustring importance then maybe they should be hidden?

File: mlfinlab/codependence/information_matrix.py [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+This script implements the creation of dependence and distance matrix for the
```
Comment (Jackal08) [inline mlfinlab/codependence/information_matrix.py]: We should elaborate on the type of distance and why plus some paper references. re our conversation on slack, I highlighted another distance metric which is found in the literature but different to this.

File: mlfinlab/codependence/information_matrix.py (hunk 13 above)
Comment (proskurin) [reply mlfinlab/codependence/information_matrix.py]: We should make a separate file for matrix calculations, the users can use it for a separate research purposes

File: mlfinlab/codependence/information_matrix.py [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    """
+
+    info_mtx = pd.DataFrame(index=df.columns)
+    features_cols = df.columns.values
+
+    algo = None
+    if base_algorithm == 'variation_of_information':
+        algo = variation_of_information_score
+    elif base_algorithm == 'mutual_information':
+        algo = mutual_info_score
```
Comment (proskurin) [inline mlfinlab/codependence/information_matrix.py]: We need get_mutual_info(de Prado) instead of mutual_info_score (scipy)
